src/value_generators/src/DocumentGenerator.cpp

```diff
@@ -1113,6 +1116,108 @@ class FastRandomStringGenerator : public StringGenerator {
+template <typename F, typename T>
+T convertToNumeric(convert<F, T>, const F& from)
+{
+    return (T)from;
+}
+
+// Special conversion for decimal
+template<typename F>
+bsoncxx::decimal128 convertToNumeric(convert<F, bsoncxx::decimal128>, const F& from) {
+    // mongocxx doesn't define nice converters from numeric types to decimal, so we convert first
+    // to string and then to decimal
+    return bsoncxx::decimal128(std::to_string(from));
+}
+
+// String conversions
+bsoncxx::decimal128 convertToNumeric(convert<std::string, bsoncxx::decimal128>,
+        const std::string& s) {
+    return bsoncxx::decimal128(s);
+}
+double convertToNumeric(convert<std::string, double>, const std::string& s) {
+    return stod(s);
+}
+int64_t convertToNumeric(convert<std::string, int64_t>, const std::string& s) {
+    return stol(s);
+}
+int32_t convertToNumeric(convert<std::string, int32_t>, const std::string& s) {
+    return stoi(s);
+}
+
+// This is the only overload we run directly, and it will route to the correct template.
+template <typename F, typename T>
+T convertToNumeric(const F& from) {
+    return convertToNumeric(convert<F, T>{}, from);
+}
+
+// overloaded struct for use with std::variants.
+template<class... Ts>
+struct overloaded : Ts... { using Ts::operator()...; };
+
+template<class... Ts>
+overloaded(Ts...) -> overloaded<Ts...>;
+
+// Generator to convert from any numeric type except decimal or string to specific numeric type.
+template<typename T>
+class NumericConversionGenerator : public Generator<T> {
+public:
+    NumericConversionGenerator(const Node& node, GeneratorArgs generatorArgs) {
+        // Try all of the possible generators for from, return when we find one that works.
+        try {
+            _generator = stringGenerator(node["from"], generatorArgs);
+            return;
+        } catch (const InvalidConversionException& e) {}
+          catch (const UnknownParserException& e) {}
+        try {
+            _generator = intGenerator(node["from"], generatorArgs);
+            return;
+        } catch (const InvalidConversionException& e) {}
+          catch (const UnknownParserException& e) {}
+        try {
+            _generator = doubleGenerator(node["from"], generatorArgs);
+            return;
+        } catch (const InvalidConversionException& e) {}
+          catch (const UnknownParserException& e) {}
+        try {
+            _generator = int32Generator(node["from"], generatorArgs);
+            return;
+        } catch (const InvalidConversionException& e) {}
+          catch (const UnknownParserException& e) {}
+        BOOST_THROW_EXCEPTION(InvalidValueGeneratorSyntax(
+            "Numeric conversion generator only supports conversions from the string, double, int, "
+            "and int32 types. The type of the \"from\" field did not match any of those types."));
+    }
+    T evaluate() override {
+        return std::visit(overloaded{
+            [&](UniqueGenerator<std::string>& gen) {
```

Comment:
I'm confused why you need the reference capture in the lambdas. Does it need `[&]` for something or would plain 'ol `[]` work?

Not sure if it matters that much tbh, but fewer places for values to potentially escape the stack seems better.

Reply:
You're right, changed to `[]`